src/composition_root.rs

```diff
@@ -1,28 +1,49 @@
 use crate::networking::{TcnApiImpl, TcnApi};
-use crate::reports_updater::{TcnMatcher, ReportsUpdater, TcnDao, Preferences, PreferencesImpl, TcnDaoImpl, TcnMatcherImpl};
+use crate::reports_updater::{TcnMatcher, ReportsUpdater, TcnDao, TcnDaoImpl, TcnMatcherImpl};
+use crate::{reporting::{memo::{MemoMapperImpl}, symptom_inputs::{SymptomInputsSubmitterImpl, SymptomInputsSubmitter}}, preferences::{Preferences, PreferencesImpl}, tcn_ext::tcn_keys::{TcnKeysImpl, TcnKeys}};
 use once_cell::sync::Lazy;
 use parking_lot::RwLock;
 
 pub struct CompositionRoot<
-PreferencesType: Preferences, TcnDaoType: TcnDao, TcnMatcherType: TcnMatcher, ApiType: TcnApi> {
+  PreferencesType: Preferences, TcnDaoType: TcnDao, TcnMatcherType: TcnMatcher, ApiType: TcnApi, 
+  // TODO don't pass concrete type for MemoMapper / TcnKeys here?
+  SymptomInputsSubmitterType: SymptomInputsSubmitter<MemoMapperImpl, TcnKeysImpl<PreferencesType>, ApiType>, 
+> {
   pub api: ApiType,
-  pub reports_updater: ReportsUpdater<PreferencesType, TcnDaoType, TcnMatcherType, ApiType>
+  pub reports_updater: ReportsUpdater<PreferencesType, TcnDaoType, TcnMatcherType, ApiType>,
+  pub symptom_inputs_submitter: SymptomInputsSubmitterType
 }
 
-pub static COMP_ROOT: Lazy<CompositionRoot<PreferencesImpl, TcnDaoImpl, TcnMatcherImpl, TcnApiImpl>> = 
+pub static COMP_ROOT: Lazy<
+  CompositionRoot<
+    PreferencesImpl, TcnDaoImpl, TcnMatcherImpl, TcnApiImpl, 
+    SymptomInputsSubmitterImpl<MemoMapperImpl, TcnKeysImpl<PreferencesImpl>, TcnApiImpl>
+  >
+> = 
   Lazy::new(|| create_comp_root());
 
-fn create_comp_root() -> CompositionRoot<PreferencesImpl, TcnDaoImpl, TcnMatcherImpl, TcnApiImpl> {
-  let api = TcnApiImpl {};
-  let preferences = PreferencesImpl { config: RwLock::new(confy::load("coepi").unwrap()) };
+fn create_comp_root() -> CompositionRoot<
+  PreferencesImpl, TcnDaoImpl, TcnMatcherImpl, TcnApiImpl, 
+  SymptomInputsSubmitterImpl<MemoMapperImpl, TcnKeysImpl<PreferencesImpl>, TcnApiImpl>
+> {
+  // FIXME pass the same instances / references
```

Comment:
https://doc.rust-lang.org/book/ch05-01-defining-structs.html#ownership-of-struct-data

Reply:
I'm merging it like this for now. Attempted to pass references / use lifetimes, but it fails with the `Preferences`, because they use `RwLock` and this uses unsafe code internally. Will fix in a follow up PR. In case you want to give it a look too, here's a self contained snippet:
```rust
struct Lock<T: ?Sized> {
  data: UnsafeCell<T>
}

struct MyPrefs<'a> {
  field1: &'a Lock<i32>
}

struct Dependency<'a> {
  field1: &'a MyPrefs<'a>
}

struct DependencyGraph<'a> {
  dep: Dependency<'a>,
}

fn create_dep_graph() -> DependencyGraph<'static> {
  let my_prefs = &MyPrefs { field1: &Lock { data: UnsafeCell::new(1) } };
  let dep = Dependency { field1: my_prefs };
  DependencyGraph { dep }
}
```

Reply:
And here the snippet again using `RwLock`
```rust
struct MyPrefs<'a> {
  field1: RwLock<&'a str>
}

struct Dependency2<'a> {
  field1: &'a MyPrefs<'a>
}

struct DependencyGraph<'a> {
  dep2: Dependency2<'a>,
}

fn create_dep_graph() -> DependencyGraph<'static> {
  let my_prefs = &MyPrefs { field1: RwLock::new("") };
  let dep2 = Dependency2 { field1: my_prefs };
  DependencyGraph { dep2 }
}
```